# Trabalho final


**Autores:**

Francisco de Assis Marinho Aguiar


Leandro Da Cruz Farias

## Importação das biblotecas

In [ ]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost

from sklearn.datasets import load_wine
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import VotingClassifier, RandomForestClassifier, BaggingClassifier, AdaBoostClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, ConfusionMatrixDisplay, SCORERS
from sklearn.metrics import average_precision_score
from operator import itemgetter

import tensorflow as tf
from tensorflow import keras
from keras.datasets import mnist, fashion_mnist
from keras.models import Sequential
from keras.layers import Flatten, Dense

### Configuração das bibliotecas

In [ ]:
pd.options.display.max_columns = None

sns.set(font_scale=1.4)
sns.set_style({'font.family': 'serif', 'fontname': 'Times New Roman'})

mpl.rcParams['figure.dpi'] = 100

## Conjunto de dados

### Dados de treino

In [ ]:
!wget -O training_dataset.csv https://drive.google.com/u/0/uc?id=1sQei4yJqUKDrB-EHdHW2ZN3hsO3EqJCS&export=download

--2021-06-25 00:06:36--  https://drive.google.com/u/0/uc?id=1sQei4yJqUKDrB-EHdHW2ZN3hsO3EqJCS
Resolving drive.google.com (drive.google.com)... 172.217.12.238, 2607:f8b0:4004:807::200e
Connecting to drive.google.com (drive.google.com)|172.217.12.238|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://doc-0s-98-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/55s6sjcqlu1igstotkqcer6ku0h8ckfd/1624579575000/12797427796612289247/*/1sQei4yJqUKDrB-EHdHW2ZN3hsO3EqJCS [following]
--2021-06-25 00:06:39--  https://doc-0s-98-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/55s6sjcqlu1igstotkqcer6ku0h8ckfd/1624579575000/12797427796612289247/*/1sQei4yJqUKDrB-EHdHW2ZN3hsO3EqJCS
Resolving doc-0s-98-docs.googleusercontent.com (doc-0s-98-docs.googleusercontent.com)... 172.217.2.97, 2607:f8b0:4004:80a::2001
Connecting to doc-0s-98-docs.googleusercontent.com (doc-0s-98-docs.googleusercontent.com)|172.217.2.97|

In [ ]:
training_df = pd.read_csv('training_dataset.csv')
training_df.head()

encounter_id  hospital_id   age        bmi  elective_surgery  ethnicity  \
0        214826          118  68.0  22.732803                 0  Caucasian   
1        246060           81  77.0  27.421875                 0  Caucasian   
2        276985          118  25.0  31.952749                 0  Caucasian   
3        262220          118  81.0  22.635548                 1  Caucasian   
4        201746           33  19.0        NaN                 0  Caucasian   

  gender  height hospital_admit_source           icu_admit_source  icu_id  \
0      M   180.3                 Floor                      Floor      92   
1      F   160.0                 Floor                      Floor      90   
2      F   172.7  Emergency Department       Accident & Emergency      93   
3      F   165.1        Operating Room  Operating Room / Recovery      92   
4      M   188.0                   NaN       Accident & Emergency      91   

  icu_stay_type      icu_type  pre_icu_los_days  readmission_status  weight  \
0         admit         CTICU          0.541667                   0    73.9   
1         admit  Med-Surg ICU          0.927778                   0    70.2   
2         admit  Med-Surg ICU          0.000694                   0    95.3   
3         admit         CTICU          0.000694                   0    61.7   
4         admit  Med-Surg ICU          0.073611                   0     NaN   

   albumin_apache  apache_2_diagnosis  apache_3j_diagnosis  \
0             2.3               113.0               502.01   
1             NaN               108.0               203.01   
2             NaN               122.0               703.03   
3             NaN               203.0              1206.03   
4             NaN               119.0               601.01   

   apache_post_operative  arf_apache  bilirubin_apache  bun_apache  \
0                      0           0               0.4        31.0   
1                      0           0               NaN         9.0   
2                      0           0               NaN         NaN   
3                      1           0               NaN         NaN   
4                      0           0               NaN         NaN   

   creatinine_apache  fio2_apache  gcs_eyes_apache  gcs_motor_apache  \
0               2.51          NaN              3.0               6.0   
1               0.56          1.0              1.0               3.0   
2                NaN          NaN              3.0               6.0   
3                NaN          0.6              4.0               6.0   
4                NaN          NaN              NaN               NaN   

   gcs_unable_apache  gcs_verbal_apache  glucose_apache  heart_rate_apache  \
0                0.0                4.0           168.0              118.0   
1                0.0                1.0           145.0              120.0   
2                0.0                5.0             NaN              102.0   
3                0.0                5.0           185.0              114.0   
4                NaN                NaN             NaN               60.0   

   hematocrit_apache  intubated_apache  map_apache  paco2_apache  \
0               27.4                 0        40.0           NaN   
1               36.9                 0        46.0          37.0   
2                NaN                 0        68.0           NaN   
3               25.9                 1        60.0          30.0   
4                NaN                 0       103.0           NaN   

   paco2_for_ph_apache  pao2_apache  ph_apache  resprate_apache  \
0                  NaN          NaN        NaN             36.0   
1                 37.0         51.0       7.45             33.0   
2                  NaN          NaN        NaN             37.0   
3                 30.0        142.0       7.39              4.0   
4                  NaN          NaN        NaN             16.0   

   sodium_apache  temp_apache  urineoutput_apache  ventilated_apache  \
0          134.0

### Dados de teste

In [ ]:
!wget -O test_dataset.csv https://drive.google.com/u/0/uc?id=1NKzd1DMTs93QK1_yoVmdnIOEReuGJcKn&export=download

--2021-06-25 00:06:42--  https://drive.google.com/u/0/uc?id=1NKzd1DMTs93QK1_yoVmdnIOEReuGJcKn
Resolving drive.google.com (drive.google.com)... 172.217.15.78, 2607:f8b0:4004:807::200e
Connecting to drive.google.com (drive.google.com)|172.217.15.78|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://doc-0k-98-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/fdq41ik5ogisjsm4d8b9s7s6c2qsgnr4/1624579575000/12797427796612289247/*/1NKzd1DMTs93QK1_yoVmdnIOEReuGJcKn [following]
--2021-06-25 00:06:43--  https://doc-0k-98-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/fdq41ik5ogisjsm4d8b9s7s6c2qsgnr4/1624579575000/12797427796612289247/*/1NKzd1DMTs93QK1_yoVmdnIOEReuGJcKn
Resolving doc-0k-98-docs.googleusercontent.com (doc-0k-98-docs.googleusercontent.com)... 172.217.2.97, 2607:f8b0:4004:80a::2001
Connecting to doc-0k-98-docs.googleusercontent.com (doc-0k-98-docs.googleusercontent.com)|172.217.2.97|:4

In [ ]:
test_df = pd.read_csv('test_dataset.csv')
test_df.head()

encounter_id  hospital_id   age        bmi  elective_surgery  ethnicity  \
0        206833           19  37.0  27.109375                 0  Caucasian   
1        273200           86  47.0  28.071089                 1  Caucasian   
2        271053           79  74.0  21.661538                 0  Caucasian   
3        221599          120  18.0  17.617188                 0  Caucasian   
4        177808           86  80.0  29.880373                 0  Caucasian   

  gender  height hospital_admit_source           icu_admit_source  icu_id  \
0      F   160.0                 Floor                      Floor     653   
1      M   164.0        Operating Room  Operating Room / Recovery    1035   
2      F   162.5      Acute Care/Floor                      Floor     366   
3      F   160.0                   NaN       Accident & Emergency    1000   
4      F   162.6  Emergency Department       Accident & Emergency    1037   

  icu_stay_type      icu_type  pre_icu_los_days  readmission_status  weight  \
0         admit  Med-Surg ICU          4.277083                   0    69.4   
1         admit     CCU-CTICU          0.340278                   0    75.5   
2         admit  Med-Surg ICU          2.694444                   0    57.2   
3         admit     CCU-CTICU          0.102083                   0    45.1   
4         admit          SICU          0.114583                   0    79.0   

   albumin_apache  apache_2_diagnosis  apache_3j_diagnosis  \
0             NaN               109.0               108.01   
1             NaN               308.0              1207.01   
2             NaN               124.0               305.02   
3             NaN               120.0               407.01   
4             NaN               301.0               403.01   

   apache_post_operative  arf_apache  bilirubin_apache  bun_apache  \
0                      0           0               NaN        30.0   
1                      1           0               NaN        12.0   
2                      0           0               NaN        24.0   
3                      0           0               NaN         NaN   
4                      0           0               NaN         8.0   

   creatinine_apache  fio2_apache  gcs_eyes_apache  gcs_motor_apache  \
0               1.56          NaN              4.0               6.0   
1               0.65          0.8              4.0               4.0   
2               0.46          NaN              4.0               6.0   
3                NaN          NaN              4.0               6.0   
4               0.65          NaN              3.0               6.0   

   gcs_unable_apache  gcs_verbal_apache  glucose_apache  heart_rate_apache  \
0                0.0                5.0            84.0              118.0   
1                0.0                1.0           173.0               93.0   
2                0.0                5.0            61.0              102.0   
3                0.0                5.0             NaN              103.0   
4                0.0                4.0           106.0              131.0   

   hematocrit_apache  intubated_apache  map_apache  paco2_apache  \
0               23.3                 0       128.0           NaN   
1               28.6                 1        51.0          45.0   
2               18.5                 0        48.0           NaN   
3                NaN                 0        54.0           NaN   
4               28.7                 0        67.0           NaN   

   paco2_for_ph_apache  pao2_apache  ph_apache  resprate_apache  \
0                  NaN          NaN        NaN              6.0   
1                 45.0         95.0       7.29             28.0   
2                  NaN          NaN        NaN             14.0   
3                  NaN          NaN        NaN             12.0   
4                  NaN          NaN        NaN             28.0   

   sodium_apache  temp_apache  urineoutput_apache  ventilated_apache  \
0          134.0

## Tratamento dos dados

In [ ]:
def categorize(df):
    df['ethnicity'] = pd.Categorical(df['ethnicity'])
    df['ethnicity'] = df['ethnicity'].cat.codes

    df['gender'] = pd.Categorical(df['gender'])
    df['gender'] = df['gender'].cat.codes

    df['hospital_admit_source'] = pd.Categorical(df['hospital_admit_source'])
    df['hospital_admit_source'] = df['hospital_admit_source'].cat.codes

    df['icu_admit_source'] = pd.Categorical(df['icu_admit_source'])
    df['icu_admit_source'] = df['icu_admit_source'].cat.codes

    df['icu_stay_type'] = pd.Categorical(df['icu_stay_type'])
    df['icu_stay_type'] = df['icu_stay_type'].cat.codes

    df['icu_type'] = pd.Categorical(df['icu_type'])
    df['icu_type'] = df['icu_type'].cat.codes

def normalize(df):
  for column in df.columns[1:]:
    df[column] = pd.to_numeric(df[column], errors='coerce')
    new_value = (df[column] - df[column].min())/(df[column].max() - df[column].min())
    df[column] = new_value

def remove_nan(df):
  
  for column in df.columns:
    if (df[column].isna().sum() > 0):
         df[column] = df[column].fillna(0)

### Treino

In [ ]:
categorize(training_df)
normalize(training_df)
remove_nan(training_df)

In [ ]:
training_df.head()

encounter_id  hospital_id       age       bmi  elective_surgery  ethnicity  \
0        214826     0.576355  0.764045  0.148912               0.0        0.5   
1        246060     0.394089  0.865169  0.237435               0.0        0.5   
2        276985     0.576355  0.280899  0.322972               0.0        0.5   
3        262220     0.576355  0.910112  0.147076               1.0        0.5   
4        201746     0.157635  0.213483  0.000000               0.0        0.5   

   gender    height  hospital_admit_source  icu_admit_source    icu_id  \
0     1.0  0.738140               0.333333               0.4  0.009718   
1     0.5  0.390478               0.333333               0.4  0.007775   
2     0.5  0.607981               0.266667               0.2  0.010690   
3     0.5  0.477822               0.600000               0.6  0.009718   
4     1.0  0.870012               0.000000               0.2  0.008746   

   icu_stay_type  icu_type  pre_icu_los_days  readmission_status    weight  \
0            0.0  0.285714          0.004501                 0.0  0.239484   
1            0.0  0.714286          0.006697                 0.0  0.214383   
2            0.0  0.714286          0.001425                 0.0  0.384668   
3            0.0  0.285714          0.001425                 0.0  0.156716   
4            0.0  0.714286          0.001840                 0.0  0.000000   

   albumin_apache  apache_2_diagnosis  apache_3j_diagnosis  \
0        0.323529            0.057971             0.228074   
1        0.000000            0.033816             0.092229   
2        0.000000            0.101449             0.319404   
3        0.000000            0.492754             0.547932   
4        0.000000            0.086957             0.273053   

   apache_post_operative  arf_apache  bilirubin_apache  bun_apache  \
0                    0.0         0.0          0.004992    0.219512   
1                    0.0         0.0          0.000000    0.040650   
2                    0.0         0.0          0.000000    0.000000   
3                    1.0         0.0          0.000000    0.000000   
4                    0.0         0.0          0.000000    0.000000   

   creatinine_apache  fio2_apache  gcs_eyes_apache  gcs_motor_apache  \
0           0.203125     0.000000         0.666667               1.0   
1           0.023897     1.000000         0.000000               0.4   
2           0.000000     0.000000         0.666667               1.0   
3           0.000000     0.493671         1.000000               1.0   
4           0.000000     0.000000         0.000000               0.0   

   gcs_unable_apache  gcs_verbal_apache  glucose_apache  heart_rate_apache  \
0                0.0               0.75        0.230481           0.594595   
1                0.0               0.00        0.189387           0.608108   
2                0.0               1.00        0.000000           0.486486   
3                0.0               1.00        0.260854           0.567568   
4                0.0               0.00        0.000000           0.202703   

   hematocrit_apache  intubated_apache  map_apache  paco2_apache  \
0           0.318182               0.0     0.00000      0.000000   
1           0.588068               0.0     0.03750      0.246753   
2           0.000000               0.0     0.17500      0.000000   
3           0.275568               1.0     0.12500      0.155844   
4           0.000000               0.0     0.39375      0.000000   

   paco2_for_ph_apache  pao2_apache  ph_apache  resprate_apache  \
0             0.000000     0.000000   0.000000         0.571429   
1             0.246753     0.042827   0.777587         0.517857   
2             0.000000     0.000000   0.000000         0.589286   
3             0.155844     0.237687   0.682267         0.000000   
4             0.000000     0.000000   0.000000         0.214286   

   sodium_apache  temp_apache  urineoutput_apache  ventilated_apache  \
0       0.414634     0

### Teste

In [ ]:
categorize(test_df)
normalize(test_df)
remove_nan(test_df)

## Definição dos parâmetros para o modelo

As variáveis **encounter_id** e **hospital_id** não foram consideradas para os dados de entrada.

### Treino

In [ ]:
x_training = training_df.loc[:,'age':'solid_tumor_with_metastasis'].to_numpy()
y_training = training_df['diabetes_mellitus'].to_numpy()

print('Amostras de treino:')
print(f' > x_train: {x_training.shape}')
print(f' > y_train: {y_training.shape}')

Amostras de treino:
 > x_train: (117141, 177)
 > y_train: (117141,)


### Teste

In [ ]:
x_test = test_df.loc[:,'age':'solid_tumor_with_metastasis'].to_numpy()

print('Amostras de teste:')
print(f' > x_test: {x_test.shape}')

Amostras de teste:
 > x_test: (13016, 177)


## Treinamento dos modelos

In [ ]:
def show_metric(model):
  score_key = 'roc_auc'
  cv = KFold(n_splits = 3, 
           shuffle = True, 
           random_state=0)

  n_scores = cross_val_score(model, 
                            x_training, y_training, 
                            scoring = score_key, 
                            cv = cv)
  
  print('Validação Cruzada\n')
  print(f'{score_key}: {n_scores}\n')
  print(f'{score_key} (média): {round(np.mean(n_scores), 4)}')
  print(f'{score_key} (desvio padrão): {round(np.std(n_scores), 4)}')

def generate_probs(model, file_name):
  predict_values = model.predict_proba(x_test)
  predict_values = predict_values[:,1]

  submission_df = test_df.copy()
  submission_df = submission_df[['encounter_id']]
  submission_df['diabetes_mellitus'] = predict_values
  submission_df.head()

 
  submission_df.to_csv(file_name, index=False)

### Regressão Logística

In [ ]:
logistic_model = LogisticRegression(penalty='l2', dual=False,
                                    tol=0.0001,
                                    C=1.0, 
                                    fit_intercept=True, 
                                    intercept_scaling=1,
                                    class_weight=None, 
                                    random_state=0, 
                                    solver='newton-cg', 
                                    max_iter=150, 
                                    multi_class='auto', 
                                    verbose=0, 
                                    warm_start=False,
                                    n_jobs=None,
                                    l1_ratio=None)

In [ ]:
logistic_model.fit(x_training, y_training)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=150,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='newton-cg', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
show_metric(logistic_model)

Validação Cruzada

roc_auc: [0.81903647 0.82184088 0.82664434 0.81825121 0.82106641]

roc_auc (média): 0.8214
roc_auc (desvio padrão): 0.0029


### Árvore de Decisão

In [ ]:
tree_model = DecisionTreeClassifier(criterion='entropy',
                                    max_depth=7,
                                    min_samples_split=2,
                                    random_state=0)

tree_model.fit(x_training, y_training)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=7, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [ ]:
show_metric(tree_model)

Validação Cruzada

roc_auc: [0.80807492 0.80614696 0.81519358 0.81177172 0.81239992]

roc_auc (média): 0.8107
roc_auc (desvio padrão): 0.0032


### Random Forest

In [ ]:
random_forest_model = RandomForestClassifier(n_estimators=30, 
                                             criterion='entropy',
                                             max_depth=10,
                                             min_samples_split=2,
                                             random_state=0)

random_forest_model.fit(x_training, y_training)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=10, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=30,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [ ]:
show_metric(random_forest_model)

Validação Cruzada

roc_auc: [0.8242281  0.82368064 0.83279344 0.82366734 0.82620784]

roc_auc (média): 0.8261
roc_auc (desvio padrão): 0.0035


### Soft Voting

In [ ]:
soft_voting_model = VotingClassifier(estimators=[('logistic_model', logistic_model),
                                                 ('tree_model', tree_model)],
                                     voting='soft')

soft_voting_model.fit(x_training, y_training)

VotingClassifier(estimators=[('logistic_model',
                              LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1,
                                                 l1_ratio=None, max_iter=150,
                                                 multi_class='auto',
                                                 n_jobs=None, penalty='l2',
                                                 random_state=0,
                                                 solver='newton-cg', tol=0.0001,
                                                 verbose=0, warm_start=False)),
                             ('tree_model',
                              DecisionTreeClassifier(ccp_alpha=0.0,
                                                     class_weight=None,
                                                     criterion='entropy',
                   

In [ ]:
show_metric(soft_voting_model)

Validação Cruzada

roc_auc: [0.82727646 0.82830347 0.83542278 0.83043433 0.83220984]

roc_auc (média): 0.8307
roc_auc (desvio padrão): 0.0029


### Bagging

In [ ]:
bagging_model = BaggingClassifier(KNeighborsClassifier(n_neighbors=3),
                                  n_estimators=800,
                                  max_samples=100)

bagging_model.fit(x_training, y_training)

BaggingClassifier(base_estimator=KNeighborsClassifier(algorithm='auto',
                                                      leaf_size=30,
                                                      metric='minkowski',
                                                      metric_params=None,
                                                      n_jobs=None,
                                                      n_neighbors=3, p=2,
                                                      weights='uniform'),
                  bootstrap=True, bootstrap_features=False, max_features=1.0,
                  max_samples=100, n_estimators=800, n_jobs=None,
                  oob_score=False, random_state=None, verbose=0,
                  warm_start=False)

In [ ]:
show_metric(bagging_model)

### Boosting

**Ada Boost com Regressão Logística**

In [ ]:
ada_model = AdaBoostClassifier(LogisticRegression(),
                               n_estimators=300, 
                               learning_rate=0.9)

ada_model.fit(x_training, y_training)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=LogisticRegression(C=1.0, class_weight=None,
                                                     dual=False,
                                                     fit_intercept=True,
                                                     intercept_scaling=1,
                                                     l1_ratio=None,
                                                     max_iter=100,
                                                     multi_class='auto',
                                                     n_jobs=None, penalty='l2',
                                                     random_state=None,
                                                     solver='lbfgs', tol=0.0001,
                                                     verbose=0,
                                                     warm_start=False),
                   learning_rate=0.9, n_estimators=300, random_state=None)

### XGBoost

In [ ]:
xgb_model = xgboost.XGBClassifier(booster='gbtree', 
                                  learning_rate=0.05,
                                  n_estimators=400,
                                  max_depth=9, 
                                  use_label_encoder=False)

xgb_model.fit(x_training, y_training)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.05, max_delta_step=0, max_depth=9,
              min_child_weight=1, missing=None, n_estimators=400, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, use_label_encoder=False, verbosity=1)

### Ensembles de ensembles

In [ ]:
soft_full_model = VotingClassifier(estimators=[('logistic_model', logistic_model),
                                                 ('tree_model', tree_model),
                                                 ('random_forest_model', random_forest_model),
                                                 ('ada_model', ada_model),
                                                 ('xgb_model', xgb_model)], 
                                     voting='soft')

soft_full_model.fit(x_training, y_training)

VotingClassifier(estimators=[('logistic_model',
                              LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1,
                                                 l1_ratio=None, max_iter=150,
                                                 multi_class='auto',
                                                 n_jobs=None, penalty='l2',
                                                 random_state=0,
                                                 solver='newton-cg', tol=0.0001,
                                                 verbose=0, warm_start=False)),
                             ('tree_model',
                              DecisionTreeClassifier(ccp_alpha=0.0,
                                                     class_weight=N...
                                            colsample_bytree=1, gamma=0,
                     

### Rede Neural

**Multilayer Perceptron (Scikit-Learn)**

In [ ]:
mlp_model = MLPClassifier(max_iter=600,
                          random_state=0,
                          verbose=0)

mlp_model.fit(x_training, y_training)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=600,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=0, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=0, warm_start=False)

In [ ]:
show_metric(mlp_model)

Validação Cruzada

roc_auc: [0.78109678 0.78319494 0.78175228]

roc_auc (média): 0.782
roc_auc (desvio padrão): 0.0009


**Multilayer Perceptron (Keras)**

In [ ]:
keras_y_training = keras.utils.to_categorical(y_training, num_classes=2)
keras_y_training

array([[0., 1.],
       [0., 1.],
       [1., 0.],
       ...,
       [0., 1.],
       [1., 0.],
       [1., 0.]], dtype=float32)

In [ ]:
mlp_keras_model = Sequential()
mlp_keras_model.add(Dense(200, input_shape=(177,), activation='relu'))
mlp_keras_model.add(Dense(100, activation='relu'))
mlp_keras_model.add(Dense(2, activation='softmax'))

mlp_keras_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 200)               35600     
_________________________________________________________________
dense_1 (Dense)              (None, 100)               20100     
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 202       
Total params: 55,902
Trainable params: 55,902
Non-trainable params: 0
_________________________________________________________________


In [ ]:
mlp_keras_model.compile(loss='categorical_crossentropy',
                  optimizer='Adam',
                  metrics=['accuracy'])

In [ ]:
callback = tf.keras.callbacks.EarlyStopping(monitor='loss', patience=3)

hist = mlp_keras_model.fit(x_training, keras_y_training, epochs=10, verbose=1, callbacks=[callback])

Epoch 1/10
3661/3661 [==============================] - 19s 2ms/step - loss: 0.4304 - accuracy: 0.8083
Epoch 2/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3936 - accuracy: 0.8227
Epoch 3/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3933 - accuracy: 0.8224
Epoch 4/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3911 - accuracy: 0.8237
Epoch 5/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3844 - accuracy: 0.8253
Epoch 6/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3838 - accuracy: 0.8262
Epoch 7/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3833 - accuracy: 0.8270
Epoch 8/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3792 - accuracy: 0.8272
Epoch 9/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.3753 - accuracy: 0.8282
Epoch 10/10
3661/3661 [==============================] - 6s 2ms/step - loss: 0.37

## Comparação entre os modelos

Modelo  | Area Under ROC Curve (média) 
-------------------|------------------
Regressão Logística | 0.8214 
Árvore de Decisão | 0.8107
Random Forest | 0.8261
Soft Voting | 0.8307
Bagging |
Boosting |
XGBoost |
Ensembles de ensembles | 0.84431
Multilayer Perceptron (Scikit-Learn) | 0.81139
Multilayer Perceptron (Keras) | 0.84906